File: angular/src/app/bills/models/no-interest-debt.ts
import {Bill} from "./bill.model";
/**
 * Created by killerkast on 7/2/17.
 */
export class NoInterestDebt extends Bill{
  private _noInterestPaymentDate;
  private _noInterestStartingBalance;

  constructor(id?: string, name?: string, description?: string,
              paymentAmount?: number, noInterestPaymentDate?: number, noInterestStartingBalance?: number){
    super(id, name, description, paymentAmount, 'interest-baring-debt');
    this._noInterestPaymentDate = noInterestPaymentDate;
    this._noInterestStartingBalance = noInterestStartingBalance;
  }

  get noInterestPaymentDate() {
    return this._noInterestPaymentDate;
  }

  set noInterestPaymentDate(value) {
    this._noInterestPaymentDate = value;
  }

  get noInterestStartingBalance() {
    return this._noInterestStartingBalance;
  }

  set noInterestStartingBalance(value) {
    this._noInterestStartingBalance = value;
  }

  updateBill(nid: NoInterestDebt){
    super.updateBill(nid);
    this.noInterestPaymentDate = nid.noInterestPaymentDate;
    this.noInterestStartingBalance = nid.noInterestStartingBalance;
  }

  getBill(){
    let noInterestDebt = super.getBill();
    noInterestDebt['paymentDate'] = this.noInterestPaymentDate;
    noInterestDebt['startingBalance'] = this.noInterestStartingBalance;
    return noInterestDebt;
  }
}

File: src/main/java/com/killerkast/entities/NoInterestDebt.java
package com.killerkast.entities;

import java.math.BigDecimal;

public class NoInterestDebt extends Bill {

    private Integer paymentDate;
    private BigDecimal startingBalance;

    public NoInterestDebt(){}

    public NoInterestDebt(String name, String description, BigDecimal paymentAmount, Integer paymentDate, BigDecimal startingBalance) {
        super(name, description, paymentAmount);
        this.paymentDate = paymentDate;
        this.startingBalance = startingBalance;
    }

    public Integer getPaymentDate() {
        return paymentDate;
    }

    public void setPaymentDate(Integer paymentDate) {
        this.paymentDate = paymentDate;
    }

    public BigDecimal getStartingBalance() {
        return startingBalance;
    }

    public void setStartingBalance(BigDecimal startingBalance) {
        this.startingBalance = startingBalance;
    }

    @Override
    public String toString() {
        return "{" +
                "No Interest Debt:{" +
                "id: " + this.getId() +
                ", name: " + this.getName() +
                ", description: " + this.getDescription() +
                ", paymentAmount: " + this.getPaymentAmount().toString() +
                ", paymentMonth: " + this.getStartingBalance().toString() +
                ", paymentDay: " + this.getPaymentDate() +
                "}}";
    }

    @Override
    public boolean equals(Object o) {
        if(!super.equals(o)) return false;
        if (this == o) return true;
        if (!(o instanceof NoInterestDebt)) return false;
        if (!super.equals(o)) return false;

        NoInterestDebt that = (NoInterestDebt) o;

        if (!getPaymentDate().equals(that.getPaymentDate())) return false;
        return getStartingBalance().equals(that.getStartingBalance());
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + getPaymentDate().hashCode();
        result = 31 * result + getStartingBalance().hashCode();
        return result;
    }
}
